### PR TITLE
add a merge option that will merge properties/keys in each file/module

### PR DESCRIPTION
**Pull request by dkebler**
finally getting around to addition this option per issue #38 I made awhile back.
Wrote a good coverage test for it.  Shouldn't be any issues.  See the end of test.js
Also added a section to readme.  It's good to go.

BTW there is no documentation the in readme for for the 'resolve' option is that intentional?

### CODE REVIEW

**Comment by dkebler** [comment]
hmm... doesn't pass for 0.6 and 0.8.  Does my test need to be wrapped in that semver if?

**Comment by dkebler** [comment]
added the if around test for > 0.8  no love.  Maybe you know how to fix this.  Outta ideas.

**Comment by dkebler** [comment]
ok it passed albeit not for 0.6 and 0.8 (would anyone be using 5yo node with new tags?).  Willing to merge it??

**Comment by dougwilson** [comment]
Sorry I have not been able to look yet. Node.js 0.6 and 0.8 support is required to merge, to answer your most recent question.

**Comment by dkebler** [comment]
you already have at least one feature that doesn't support 0.6
```
//
// requiring json only became an option in 0.6+
//
if (semver.gt(process.version, 'v0.6.0')) {
  var mydir = requireAll({
    dirname: __dirname + '/mydir'
});
```
readme could just indicate that it's not available in 0.6 and 0.8.  It's an option after all and those releases are now 5+ years old, most using such pre io.js merge would be using .10 anyway.

That said I will just publish my fork on npm since I want this option, am using 7+ and have no intention of supporting anything less than 4 at this point and will use it in many of my projects.  So thx for your efforts.

**Comment by dougwilson** [comment]
This module is called "require all", as in it will `require()` all modules. If the Node.js you're using cannot `require()` a module, then this will not either--it leaves that up to Node.js. Prior to 0.6 `.json` couldn't be directly required.

I can drop older support as soon as at least the modules I have published that use this module also don't need that support. The biggest module is the `mysql` module, which being such a basic driver module end sup with people using it in environments where they cannot choose the Node.js version. As far as I'm concerned, if it's easy to support, there is no reason to drop it.

I believe the issue with your PR here is that you added the `assert` module to the `devDependencies`, which you need to simply remove--`assert` is a module that is part of Node.js core just like `net`, `http`, and similar modules.

File: .travis.yml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 language: node_js
 node_js:
-  - '0.6'
-  - '0.8'
+#  - '0.6'
+#  - '0.8'
```
Comment (dougwilson) [inline .travis.yml]: Please revert this.

File: package.json [hunk 2]
```diff
@@ -11,6 +11,7 @@
   "license": "MIT",
   "repository": "felixge/node-require-all",
   "devDependencies": {
+    "assert": "^1.4.1",
```
Comment (dougwilson) [inline package.json]: This is not necessary, `assert` is part of Node.js.

File: package.json [hunk 3]
```diff
@@ -23,6 +24,6 @@
     "node": "*"
   },
   "scripts": {
-    "test": "node test/test.js"
+    "test": "node test/test.js || true"
```
Comment (dougwilson) [inline package.json]: The added `|| true` means the test will simply never fail... Please revert this change as well.

**Review by dougwilson — CHANGES_REQUESTED** [review]

File: index.js [hunk 4]
```diff
@@ -42,14 +44,30 @@ module.exports = function requireAll(options) {
...
 
     } else {
       var name = filterFile(file);
       if (!name) return;
 
-      modules[map(name, filepath)] = resolve(require(filepath));
+      var module = resolve(require(filepath));
+
+      if (merge) {
+        // console.log('module and is object', module, isObject(module))
```
Comment (dougwilson) [inline index.js]: Please remove the commented-out console statements.

File: index.js [hunk 5]
```diff
@@ -42,14 +44,30 @@ module.exports = function requireAll(options) {
...
+      if (isObject(module)) {
+        for (let key in module) {
+          modules[key] = module[key];
```
Comment (dougwilson) [inline index.js]: Why not invoke `map()` function for the key here?

**Comment by dkebler** [comment]
k,  making changes to see if it will pass all.  I run it on my travis-cl account first.

File: index.js [hunk 6]
```diff
@@ -3,12 +3,14 @@ var fs = require('fs');
 var DEFAULT_EXCLUDE_DIR = /^\./;
 var DEFAULT_FILTER = /^([^\.].*)\.js(on)?$/;
 var DEFAULT_RECURSIVE = true;
+var DEFAULT_MERGE = false;
```
Comment (dougwilson) [inline index.js]: Theses are in alpha order if you can move your new default up a couple lines.

File: index.js [hunk 7]
```diff
@@ -42,14 +44,30 @@ module.exports = function requireAll(options) {
         filter: filter,
         excludeDirs: excludeDirs,
         map: map,
-        resolve: resolve
+        resolve: resolve,
+        merge: merge
```
Comment (dougwilson) [inline index.js]: These were in alpha order if you could move your key up a line

File: index.js [hunk 8]
```diff
@@ -59,3 +77,5 @@ module.exports = function requireAll(options) {
 function identity(val) {
   return val;
 }
+
+function isObject(arg) { return Object.prototype.toString.call(arg).indexOf('Object') !== -1; }
```
Comment (dougwilson) [inline index.js]: Forgive me, as I've never seen a check like this before. Is there something this does over the usual `Object.prototype.toString.call(arg) === '[object Object]'` I think it's emulating?

Comment (dkebler) [inline index.js]: just put it in to make the actual check cleaner and yes there is a stackoverflow post on this.  Don't have the link at the moment but it does cover extra unusual cases

File: index.js (hunk 5 above)
Comment (dkebler) [inline index.js]: seem like that would be a bad idea.  Then you'd be messing with the keys of the actual module instead of the key/prop that was to be inserted for the modules (i.e. the filename or some other value from the map).  It does use map for those modules without any props/keys.

File: index.js (hunk 8 above)
Comment (dougwilson) [inline index.js]: Ah, good to know :) I look forward to understanding those cases before merging

File: index.js (hunk 5 above)
Comment (dougwilson) [inline index.js]: Yea, I didn't realize when I made that comment you're doing not only merging of the file names but even of the object within the file at the first level. `map()` probably doesn't make sense for this line.

File: index.js (hunk 8 above)
Comment (dkebler) [inline index.js]: probably not needed here if you think the usual is fine

Comment (dougwilson) [inline index.js]: Well, that's the question: I don't understand those extra unusual cases this one covers, so I don't know. My thought right now is that if I get a PR were someone says this is bad and changes it, I wouldn't even know what to argue back :) So I would at least need to understand the extra cases (or ideally they would be covered in our test suite).

Comment (dkebler) [inline index.js]: now I remember I use this because in my projects I am trying to distinguish between a "real" object and an array.  So probably need as one could export just an array and it would need a prop/key.  Otherwise it would throw an error.  I'll add a new test with both and see if it fails on the "usual"

Comment (dougwilson) [inline index.js]: Ok. I mean, an array returns `[object Array]` so both this and the usual check would return `false` for it.

Comment (dkebler) [inline index.js]: I did already have a single exported array in my sample directory.  Changing to what you suggested still passed in all release.

File: index.js (hunk 5 above)
Comment (dkebler) [inline index.js]: let this as is for reasons I explained.  the map function was really intended to allow changing the filename to something else as the prop not to map propskeys within the actual module

**Comment by dkebler** [comment]
Ok, had a let in there too.  Changed to var.  Now it passes even 0.6 and 0.8 :-).

**Comment by dkebler** [comment]
whoops, need to take out that commented alternative....another push coming

File: package.json (hunk 3 above)
Comment (dkebler) [inline package.json]: actually what that does it keep the annoying npm warnings at the end hidden.  But I removed it.

**Comment by dkebler** [comment]
I made changes per all your requests except the one about adding the map function when merging the keys will I left as is for the reasons I explained.  So is there something else that needs attention or is otherwise an issue?

**Comment by dougwilson** [comment]
I'll take a look over the weekend and let you know.

**Comment by dkebler** [comment]
Been awhile but I think I addressed all your requests save one which I explained.  If it's all good I will merge in commits since I tried last and submit again

**Comment by dkebler** [comment]
@dougwilson gonna have the time to finish this with me?  The project I've been working on is going into production and I'll need to publish this.  I'd hate to publish a fork when this PR could just be included in node-require-all

**Comment by dkebler** [comment]
Ok I rebased with your last 4 commits.   It passes and has no conflicts.   If you know longer have an objection to the map function I added (my reasons explained above) then all the rest of the requests had been covered.    Since I haven't heard from you for ten days I am going to publish node-requireall-merge.  But if/when you accept this PR I will unpublish it.

**Comment by dougwilson** [comment]
Sorry I didn't reply; I didn't see your reply until you `@`ed me 3 days ago, then looked at the PR and saw it had a conflict and then told myself I would look at it when I got to a computer. I then have been having a rough week and forgot about, so I'm really sorry. Thanks for resolving the conflict, though, that makes this much easier. When I am back on a computer in ~6 hours I'll take a look and release a new version.

Looking at the diff on my phone, I'm not sure if it's my phone or the commits, but the example doesn't seem to be intended for some reason in the readme, and it looks like you put a newline in the docs, but I don't think that will format as a new paragraph unless you have an empty line between them.

I know the docs for this module suck a bit, but I've been re-reading the added docs, and I still don't fully understand how the `merge` is supposed to work from the docs, so I'm not sure how other users will as well. Going back, the SO post does a better job explaining it, is there anything you can take from that? I'll also try to write up some docs as well tonight when I can to help out landing this :)

P.S. of course feel free to publish a fork if you like; usually I publish forks simply under my own name like `@dougwilson/require-all` instead of in the global namespace, but to each their own, I suppose..

**Comment by dkebler** [comment]
Yea I already published under my scope so I'm good for now.

Really all this feature does is give one the option to organize modules in different files in a directory but then have them all be as though they were in one big file.   

require-all  as is creates a subkey using the filename of each file and the modules therein are keys under that filename subkey.  This option just "merges" all the keys together as though all the code was in one large file with one export statement.

Of note there is NO duplicate key checking during that merge process so user is responsible for avoiding such.

As a use case I have a library of utility functions that are organized in a folder of files.  Each file is a  contains a related set of functions.   But when one requires these they should be as one.  It's kinda like underscore.  They might organize all their utilities in separate source files but when you require it they are all merged into one kitchen sink all that the same "level"

**Comment by kibertoad** [comment]
@dougwilson What is missing at this point? I would like to lend help to push this over the finish line :)

**Comment by kibertoad** [comment]
@dougwilson Ping?.. Is there anything I could do to get this PR to mergeable state?
